engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterRunLoop.swift

```diff
@@ -0,0 +1,159 @@
+      }
+    )
+    guard let createdSource = CFRunLoopSourceCreate(kCFAllocatorDefault, 0, &sourceContext) else {
+      fatalError("Failed to create CFRunLoopSource")
+    }
+    source = createdSource
+    CFRunLoopAddSource(runLoop, source, .commonModes)
+    CFRunLoopAddSource(runLoop, source, Self.flutterRunLoopMode)
+
+    var timerContext = CFRunLoopTimerContext(
+      version: 0,
+      info: Unmanaged.passUnretained(self).toOpaque(),
+      retain: nil,
+      release: nil,
+      copyDescription: nil
+    )
+    guard
+      let createdTimer = CFRunLoopTimerCreate(
+        kCFAllocatorDefault,
+        CFAbsoluteTime.greatestFiniteMagnitude,
+        CFAbsoluteTime.greatestFiniteMagnitude,
+        0,
+        0,
+        { timer, info in
+          let runner = Unmanaged<FlutterRunLoop>.fromOpaque(info!).takeUnretainedValue()
+          runner.performExpiredTasks()
+        },
+        &timerContext
+      )
+    else {
+      fatalError("Failed to create CFRunLoopTimer")
+    }
+    timer = createdTimer
+    CFRunLoopAddTimer(runLoop, timer, .commonModes)
+    CFRunLoopAddTimer(runLoop, timer, Self.flutterRunLoopMode)
+  }
+
+  deinit {
+    CFRunLoopTimerInvalidate(timer)
+    CFRunLoopRemoveTimer(runLoop, timer, .commonModes)
+    CFRunLoopRemoveTimer(runLoop, timer, Self.flutterRunLoopMode)
+    CFRunLoopSourceInvalidate(source)
+    CFRunLoopRemoveSource(runLoop, source, .commonModes)
+    CFRunLoopRemoveSource(runLoop, source, Self.flutterRunLoopMode)
+  }
+
+  // Ensures that the `FlutterRunLoop` for main thread is initialized. Only
+  // needs to be called once and must be called on the main thread.
+  @objc public static func ensureMainLoopInitialized() {
+    assert(Thread.isMainThread, "Must be called on the main thread.")
+    if _mainRunLoop == nil {
+      _mainRunLoop = FlutterRunLoop()
+    }
+  }
+
+  // The `FlutterRunLoop` for the main thread.
+  @objc public static var mainRunLoop: FlutterRunLoop {
+    assert(
+      _mainRunLoop != nil,
+      "Main run loop has not been initialized. Call ensureMainLoopInitialized() first."
+    )
+    return _mainRunLoop!
+  }
+
+  // Schedules a block to be executed on the main thread.
+  @objc(performBlock:afterDelay:)
+  public func perform(block: @escaping () -> Void, afterDelay delay: TimeInterval) {
+    tasksLock.lock()
+    defer { tasksLock.unlock() }
+
+    tasks.append(Task(block: block, targetTime: CFAbsoluteTimeGetCurrent() + delay))
+    if delay > 0 {
+      rearmTimer()
+    } else {
+      CFRunLoopSourceSignal(source)
+      CFRunLoopWakeUp(runLoop)
+    }
+  }
+
+  // Schedules a block to be executed on the main thread after a delay.
+  @objc(performBlock:)
+  public func perform(block: @escaping () -> Void) {
+    perform(block: block, afterDelay: 0)
+  }
+
+  private func performExpiredTasks() {
+    var expiredTasks: [Task] = []
+
+    tasksLock.lock()
+    let now = CFAbsoluteTimeGetCurrent()
+    tasks.removeAll { task -> Bool in
```

Comment:
nit: try to avoid side effect in swift functional programming (i.e. prefer pure functions)

I'd just pass the loop twice: 

```
let expired = tasks.filter { $0.expired }
let remaining = tasks.filter { !$0.expired }
```
Or if you wanna do one pass, can do a reduce

```

let (expired, remaining) = tasks.reduce(into: (expired: [Task](), remaining:[Task]())) { partialResult, task in
  if task.expired {
    partialResult.expired += [task]
  } else {
    partialResult.remaining += [task]
  }
}
```

Reply:
Funny I had originally written it that way, then rewrote it this way, thinking this was more Swifty 😂 Thanks!

Reply:
Done. Simple, fast, and least instructions: plain for loop.

Reply:
I guess a plain for loop is fine if you'd like. Many people prefer functional programming since it's less error-prone (stateless, immutable).